evm/chains/tester/__init__.py

```diff
@@ -42,91 +48,76 @@ def create_header_from_parent(cls,
+        raise ValidationError(
+            "The `dao_start_block` parameter is only valid for the 'homestead' "
+            "fork rules.  The 'homestead' VM was not included in the provided "
+            "fork configuration"
+        )
 
-        remaining_blocks_not_declared = (
-            homestead_start_block is None and
-            tangerine_whistle_start_block is None
+    # Validate that there is at least one start block for block 0
+    start_blocks = sorted(set(start_block for start_block, _ in fork_start_blocks))
+    if 0 not in start_blocks:
+        raise ValidationError(
+            "At least one VM must start at block 0.  Got start blocks: "
+            "{0}".format(start_blocks)
         )
-        if spurious_dragon_block > 0 and remaining_blocks_not_declared:
-            yield (0, TangerineWhistleTesterVM)
-
-    if tangerine_whistle_start_block is not None:
-        yield (tangerine_whistle_start_block, TangerineWhistleTesterVM)
-
-        # If the EIP150 rules do not start at block 0 and homestead_start_block has not
-        # been configured for a specific block, configure homestead_start_block to start at
-        # block 0.
-        if tangerine_whistle_start_block > 0 and homestead_start_block is None:
-            HomesteadTesterVM = BaseHomesteadTesterVM.configure(
-                dao_fork_block_number=0,
-            )
-            yield (0, HomesteadTesterVM)
-
-    if homestead_start_block is not None:
-        if dao_start_block is False:
-            # If dao_start_block support has explicitely been configured as `False` then
-            # mark the HomesteadTesterVM as not supporting the fork.
-            HomesteadTesterVM = BaseHomesteadTesterVM.configure(support_dao_fork=False)
-        elif dao_start_block is not None:
-            # Otherwise, if a specific dao_start_block fork block has been set, use it.
-            HomesteadTesterVM = BaseHomesteadTesterVM.configure(
-                dao_fork_block_number=dao_start_block,
-            )
-        else:
-            # Otherwise, default to the homestead_start_block block as the
-            # start of the dao_start_block fork.
-            HomesteadTesterVM = BaseHomesteadTesterVM.configure(
-                dao_fork_block_number=homestead_start_block,
-            )
-        yield (homestead_start_block, HomesteadTesterVM)
 
-        # If the homestead_start_block block is configured to start after block 0, set the
-        # frontier rules to start at block 0.
-        if homestead_start_block > 0:
-            yield (0, FrontierTesterVM)
+    ordered_fork_start_blocks = sorted(fork_start_blocks, key=operator.itemgetter(0))
+
+    for start_block, fork in ordered_fork_start_blocks:
+        if isinstance(fork, type) and issubclass(fork, BaseVM):
+            vm_class = fork
+        elif isinstance(fork, str):
+            vm_class = MAINNET_VMS[fork]
+        else:
+            raise Exception("Invariant: unreachable code path")
```

Comment:
Do the type definitions make this literally unreachable? If it's possible for someone to supply a bad fork value here, then a better error message would be nice.

Reply:
I believe that the typing makes this unreachable (with code that has passed the `mypy` checks) but I still like having these in place to basically document the remaining logic branch.